### PR TITLE
Extend questionnaires for supplies reports

### PR DESCRIPTION
**Pull request by qptr**
**Story card:** [sc-10324](https://app.shortcut.com/simpledotorg/story/10324)

## Because

Facilities need to send reports of available supplies to aid resource allocation. This is currently done manually in a lot of facilities across countries. This PR extends the existing questionnaire model to create supplies reports.

## This addresses

- Adds form for supplies report
- Adds different forms for Ethiopia, Sri Lanka and Bangladesh
- Extends the questionnaire DSL to incorporate more form elements
- Sets up a regular task to pre-fill metadata for supplies reports

## Test instructions

- Checkout the new swagger docs on https://simple-review-pr-5075.herokuapp.com/api-docs
- Try to fill and submit a supplies report from the Simple app.
- Use the string inputs and radio buttons
- Edit an older form

### CODE REVIEW

File: public/documents/upload_facilities.csv [hunk 1]
```diff
@@ -1 +1 @@
-organization,facility_group,facility_name,facility_short_name,facility_type,street_address (optional),village_or_colony (optional),zone_or_block,district,pin (optional),latitude (optional),longitude (optional),size (optional),enable_diabetes_management (true/false),enable_monthly_screening_reports (true/false),enable_monthly_supplies_reports (true/false)
```
Comment (govindkrjoshi) [inline public/documents/upload_facilities.csv:1]: Unrelated change?

File: spec/controllers/api/v4/questionnaires_controller_spec.rb [hunk 2]
```diff
@@ -135,7 +135,9 @@ def create_record_list(n, options = {})
     end
 
     it "returns questionnaires with the same major dsl_version uptil given minor version" do
-      questionnaire_1 = create(:questionnaire, :active, questionnaire_type: @questionnaire_types.first, dsl_version: "1.0")
+      allow_any_instance_of(Questionnaire).to receive(:validate_layout)
```
Comment (govindkrjoshi) [inline spec/controllers/api/v4/questionnaires_controller_spec.rb:138]: What is this line testing?

Comment (govindkrjoshi) [reply spec/controllers/api/v4/questionnaires_controller_spec.rb:138]: Looks like this line is present in most of the tests in the PR changes. Are we trying to stub `validate_layout`? Curious why is that necessary.

**Review by govindkrjoshi — CHANGES_REQUESTED** [review]

File: app/models/questionnaire.rb [hunk 3]
```diff
@@ -49,10 +50,15 @@ def transform_layout(&blk)
   private
 
   def layout_schema
-    # TODO: When dsl_version is incremented, insert a switch here.
-    Api::V4::Models::Questionnaires::Version1.view_group.merge(
-      definitions: Api::V4::Schema.all_definitions
-    )
+    model = case dsl_version
```
[the remark this answers is not in the document]
Comment (govindkrjoshi) [reply app/models/questionnaire.rb:53]: If you use Hash#fetch, it should also throw an error for missing keys.

File: config/locales/api/am_ET.yml [hunk 4]
```diff
@@ -446,12 +446,13 @@ am-ET:
         cohort_subtitle: "%{number_of_patients} %{patient_or_patients} በ %{registration_period} የተመዘገቡ"
         cohort_bar_tooltip: "%{total_patients} %{patient_or_patients} የ %{threshold}"
   questionnaire_layout:
+    empty: ""
```
[the remark this answers is not in the document]
Comment (qptr) [reply config/locales/api/am_ET.yml:449]: The comment view has placeholder text that needs to be empty. Strings have a required text field. Actually @olttwa can we just set this to be "" in the layouts?

File: config/locales/api/bn_IN.yml [hunk 5]
```diff
@@ -446,12 +446,13 @@ bn-IN:
...
         cohort_bar_tooltip: "%{threshold} থাকা %{total_patients} %{patient_or_patients}"
   questionnaire_layout:
+    empty: ""
     male: "পুরুষ"
     female: "মহিলা"
     transgender: "ট্রান্সজেন্ডার"
     total: "মোট"
-    yes: "হ্যাঁ"
-    no: "না"
+    yes_lc: "হ্যাঁ"
```
[the remark this answers is not in the document]
Comment (qptr) [reply config/locales/api/bn_IN.yml:454]: @olttwa Shall we set these to `positive` and `negative`?

File: app/services/questionnaire_responses/initialize_monthly_supplies_reports.rb [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+    reports_exist = false
+    questionnaire = latest_active_supplies_reports_questionnaire
+
+    facility_ids.each do |facility_id|
+      if monthly_supplies_report_exists?(facility_id, month_date_str)
+        reports_exist = true
+      else
+        QuestionnaireResponse.create!(
+          questionnaire_id: questionnaire.id,
```
Comment (qptr) [inline app/services/questionnaire_responses/initialize_monthly_supplies_reports.rb:17]: questionnaire may not exist, we should check for this.

File: app/schema/api/v4/models/questionnaires/specimen_layout.rb [hunk 7]
```diff
@@ -0,0 +1,216 @@
+class Api::V4::Models::Questionnaires::SpecimenLayout
+  class << self
+    def dsl_version1
+      {
```
[the remark this answers is not in the document]
Comment (qptr) [reply app/schema/api/v4/models/questionnaires/specimen_layout.rb:4]: For reference, is swagger loading from a data file any where else in the project?

File: config/locales/api/am_ET.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (qptr) [reply config/locales/api/am_ET.yml:449]: Discussed with @olttwa. The translator runs over each text field inside the layout and fails if it cannot find the key in the translation file.

File: config/locales/api/bn_BD.yml [hunk 8]
```diff
@@ -1,4 +1,4 @@
-bn_BD:
+bn-BD:
```
[the remark this answers is not in the document]
Comment (olttwa) [reply config/locales/api/bn_BD.yml:1]: Bangla translations have been broken for a while. Transifex changes `-` to `_` and it didn't get caught during one of the translations.
This change fixes the translations.

For more context, refer to [this documentation in the wiki](https://github.com/simpledotorg/simple-server/blob/master/doc/wiki/transifex-localization.md#3-check-and-fix-all-top-level-language-keys)

File: spec/controllers/api/v4/questionnaires_controller_spec.rb (hunk 2 above)
Comment (olttwa) [reply spec/controllers/api/v4/questionnaires_controller_spec.rb:138]: `validate_layout` only allows DSL versions `1` and `1.1`. We're mocking it's implementation whenever we need to add more DSL versions in DB for testing purpose.

`validate_layout` is mocked twice in controller and once in model test.

File: app/schema/api/v4/models/questionnaires/specimen_layout.rb (hunk 7 above)
[the remark this answers is not in the document]
Comment (olttwa) [reply app/schema/api/v4/models/questionnaires/specimen_layout.rb:4]: We couldn't find a way to do this which is neater than current implementation. Loading from JSON file doesn't format the response on swagger UI correctly.
image
image

File: app/models/questionnaire.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (olttwa) [reply app/models/questionnaire.rb]: done.

File: app/services/questionnaire_responses/initialize_monthly_supplies_reports.rb [hunk 9]
```diff
@@ -0,0 +1,50 @@
+class QuestionnaireResponses::InitializeMonthlySuppliesReports
```
[the remark this answers is not in the document]
Comment (olttwa) [reply app/services/questionnaire_responses/initialize_monthly_supplies_reports.rb]: Moved to `Klass.new.call` interface

File: app/services/questionnaire_responses/initialize_monthly_supplies_reports.rb [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+          facility_id: facility_id,
+          content: supplies_report_content(month_date_str),
+          device_created_at: month_date,
+          device_updated_at: month_date
+        )
+      end
+    end
+
+    if reports_exist
+      Rails.logger.error("Some/all monthly supplies reports already existed during initialization task for month: %s" % month_date_str)
```
[the remark this answers is not in the document]
Comment (olttwa) [reply app/services/questionnaire_responses/initialize_monthly_supplies_reports.rb]: removed the logging & the `reports_exist` flag

File: app/services/questionnaire_responses/initialize_monthly_supplies_reports.rb [hunk 11]
```diff
@@ -0,0 +1,50 @@
+class QuestionnaireResponses::InitializeMonthlySuppliesReports
+  def self.call(date = 1.month.ago)
+    return unless Flipper.enabled?(:monthly_supplies_reports)
+
+    month_date = date.beginning_of_month
+    month_date_str = month_date.strftime("%Y-%m-%d")
```
[the remark this answers is not in the document]
Comment (olttwa) [reply app/services/questionnaire_responses/initialize_monthly_supplies_reports.rb]: In server, we send a standard date string format. Mobile displays differently only for Ethiopia. For example, `20 May 2023` gets displayed as `15-07-2015`

File: app/services/questionnaire_responses/pre_fill_monthly_screening_reports.rb [hunk 12]
```diff
@@ -1,4 +1,4 @@
-class PreFillMonthlyScreeningReports
+class QuestionnaireResponses::PreFillMonthlyScreeningReports
```
[the remark this answers is not in the document]
Comment (olttwa) [reply app/services/questionnaire_responses/pre_fill_monthly_screening_reports.rb]: done.

File: app/services/questionnaire_responses/initialize_monthly_supplies_reports.rb [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+    questionnaire = latest_active_supplies_reports_questionnaire
+
+    facility_ids.each do |facility_id|
+      if monthly_supplies_report_exists?(facility_id, month_date_str)
+        reports_exist = true
+      else
+        QuestionnaireResponse.create!(
+          questionnaire_id: questionnaire.id,
+          facility_id: facility_id,
+          content: supplies_report_content(month_date_str),
```
[the remark this answers is not in the document]
Comment (olttwa) [reply app/services/questionnaire_responses/initialize_monthly_supplies_reports.rb]: refactored.

File: public/documents/upload_facilities.csv (hunk 1 above)
Comment (olttwa) [reply public/documents/upload_facilities.csv:1]: we're following practice to have a new line at bottom of file.

File: app/services/questionnaire_responses/initialize_monthly_supplies_reports.rb (hunk 6 above)
Comment (olttwa) [reply app/services/questionnaire_responses/initialize_monthly_supplies_reports.rb]: done.

File: config/locales/api/bn_IN.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (olttwa) [reply config/locales/api/bn_IN.yml:454]: We're sticking with `_lc` & `_uc` because we're using that across other keys too in monthly reports keys.

File: app/schema/api/v4/models/questionnaires/specimen_layout.rb (hunk 7 above)
[the remark this answers is not in the document]
Comment (govindkrjoshi) [reply app/schema/api/v4/models/questionnaires/specimen_layout.rb:7]: Okay, but JSON.parse would return a hash anyways though. Or yaml.parse would work too. Am I missing something?

Don't want to block the PR for this, though.

File: lib/seed/questionnaire_seeder.rb [hunk 14]
```diff
@@ -1,16 +1,210 @@
...
+            view_type: "sub_header"
+          },
+          {
+            type: "group",
+            view_type: "input_view_group",
+            item: [
+              {
+                type: "string",
+                link_id: "monthly_supplies_report.comments",
+                text: "questionnaire_layout.empty",
```
Comment (govindkrjoshi) [inline lib/seed/questionnaire_seeder.rb:198]: Ideally we should be able to insert arbitrary strings here, not just keys in the translation files. We are coupling questionnaire to the translations. Can be fixed in the future.

File: config/locales/api/am_ET.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (govindkrjoshi) [reply config/locales/api/am_ET.yml:449]: https://github.com/simpledotorg/simple-server/pull/5107/files#r1198821126

**Review by govindkrjoshi — APPROVED** [review]
